crates/holochain_types/src/db_cache.rs

```diff
@@ -0,0 +1,575 @@
+fn update_activity(
+    activity: &mut HashMap<Arc<AgentPubKey>, ActivityState>,
+    agent: &AgentPubKey,
+    new_bounds: &ActivityBounds,
+) -> DbCacheResult<()> {
+    let prev_state = activity.get_mut(agent);
+    if !update_activity_check(prev_state.as_deref().map(|s| &s.bounds), new_bounds) {
+        return Err(DbCacheError::ActivityOutOfOrder(
+            prev_state.and_then(|p| p.integrated).unwrap_or(0),
+            new_bounds.integrated.unwrap_or(0),
+        ));
+    }
+    match prev_state {
+        Some(prev_bounds) => update_activity_inner(prev_bounds, new_bounds),
+        None => {
+            // If the new bounds have `ready_to_integrate` and do not equal zero
+            // then they are out of order.
+            if new_bounds.ready_to_integrate.map_or(false, |i| i != 0) {
+                activity.insert(
+                    Arc::new(agent.clone()),
+                    ActivityState {
+                        bounds: ActivityBounds {
+                            integrated: new_bounds.integrated,
+                            ..Default::default()
+                        },
+                        out_of_order: new_bounds.ready_to_integrate.into_iter().collect(),
+                    },
+                );
+            } else {
+                activity.insert(
+                    Arc::new(agent.clone()),
+                    ActivityState {
+                        bounds: *new_bounds,
+                        ..Default::default()
+                    },
+                );
+            }
+        }
+    }
+    DbCacheResult::Ok(())
+}
+
+fn update_activity_inner(prev_state: &mut ActivityState, new_bounds: &ActivityBounds) {
+    if new_bounds.integrated.is_some() {
+        prev_state.bounds.integrated = new_bounds.integrated;
+    }
+    update_ready_to_integrate(prev_state, new_bounds.ready_to_integrate);
+}
+
+/// Updates the ready to integrate state of an activity.
+/// This function is a bit complex but is heavily tested and maintains the
+/// chain activity can only be set to ready if it makes sense to.
+fn update_ready_to_integrate(prev_state: &mut ActivityState, new_ready: Option<u32>) {
+    if let Some(new_ready) = new_ready {
+        match prev_state {
+            ActivityState {
+                bounds:
+                    ActivityBounds {
+                        integrated: None,
+                        ready_to_integrate: ready @ None,
+                    },
+                out_of_order,
+            } => {
+                // (0) -> Ready(0)
+                if new_ready == 0 {
+                    *ready = Some(new_ready);
+                // (x) -> Out(x) where x > 0
+                } else {
+                    out_of_order.push(new_ready);
+                    out_of_order.sort_unstable();
+                }
+            }
+            // (Ready(x), a) -> (Ready(x), Out(y) where a != x')
+            // (Ready(x), Out(y..), a) -> (Ready(x), Out(y, a)) where a != x'
```

Comment:
It would be helpful to understand your notation here, for me and future readers, can you say a bit about it?